[4-6단계 - 방탈출 사용자 예약] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 이스트! 다시 만나게 되어 반갑습니다!

### ❗️프로그램 사용방법
요구사항에 없었지만 회원가입 기능을 구현하였습니다.
**ADMIN** 권한으로 사용하기 위해서는 회원가입시 `name`을 `어드민` 또는 `admin` 으로 입력하여 가입하면 어드민 권한으로 사용할 수 있습니다.
이외의 다른 이름은 모두 사용자 권한으로 가입 됩니다!

step2 리뷰도 잘 부탁드립니다🙇‍♂️

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

비슷한 메서드가 존재하고, 원시값 포장, 객체의 책임과 역할 분리에서 미흡한 부분이 있다고 생각하여 3점을 주었습니다.


## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### Service 로직과 dao 로직 판단
서비스 계층에서 처리해야할 로직과 repository 계층에서 처리해야할 로직을 판단하는 것이 어렵게 느껴졌습니다. 
계층별로 수행해야할 로직이 이스트의 기준에서 적절히 수행되고 있는지, 부족한 부분이 있다면 피드백 해주시면 좋을것 같습니다!

### 인증과 인가
인증 인가에 그리고 토큰에 대한 지식이 미흡하여 제대로 사용하고 있는지 의심이 들었습니다.
인증과 인가 부분을 사용하는 코드에서 보충하면 좋을 부분과, 학습하면 좋을 키워드를 중점으로 피드백 해주시면 좋을 것 같습니다!

### 객체들간의 책임 분리
미션을 구현하면서 비슷한 책임을 가지는 객체들이 있다고 느껴졌습니다. 각각의 도메인과 계층간의 책임과 역할이 적절히 나누어졌는지, 부족한 부분이 있는지 피드백 해주시면 좋을 것 같습니다!

이외에 특정 코드에 대한 부분은 코멘트로 남기겠습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

dye0p committed:
docs: 4단계 요구사항 업데이트

dye0p committed:
feat: 로그인 페이지 응답 매핑 구현

- GET /login 요청시 로그인 페이지를 응답한다.

dye0p committed:
feat: 회원가입 페이지 응답 매핑 구현

dye0p committed:
feat: 회원 테이블 생성

dye0p committed:
feat: 회원가입 API 구현

dye0p committed:
chore: 회원 테이블 email 컬럼 unique key 설정

dye0p committed:
feat: 회원 role 필드 추가

- 회원의 권한을 기본설정 하기 위해서 정적 팩터리 메서드을 통해서 객체를 생성하도록 변경

dye0p committed:
feat: 로그인 API 구현

- 로그인 인증에 성공하면 토큰을 발급하여 쿠키에 추가하여 응답한다.

dye0p committed:
feat: 로그인 후 사용자의 정보를 조회하는 API 구현

dye0p committed:
feat: 로그아웃 API 구현

- 로그아웃을 하면 쿠키가 삭제된다.

dye0p committed:
refactor: 쿠키와 토큰 사용 로직 클래스 분리

dye0p committed:
docs: 5단계 요구사항 업데이트

dye0p committed:
refactor: 컨트롤러에 요청된 회원정보를 객체로 바인딩 하여 컨트롤러 매개변수에 주입하도록 변경

- 요청시 Cookie에 포함된 member 정보를 조회하는 로직이 매 요청마다 중복되는 이유로 HandlerMethodArgumentResolver를 이용하여 컨트롤러에 진입하기 전에 member 객체로 바인딩하여 컨트롤러 메서드에서 주입받아서 사용할 수 있도록 함

dye0p committed:
refactor: 로그인한 사용자가 예약 생성할 수 있도록 변경

- 로그인한 사용자만 예약할 수 있도록 변경
- 사용자 예약 생성시 이름을 직접입력하지 않고, 로그인한 사용자 정보를 활용하여 예약을 생성함

dye0p committed:
refactor: 관리자가 예약 생성 시, 유저를 조회하여 선택 후 예약을 생성할 수 있도록 리팩터링

dye0p committed:
docs: 6단계 요구사항 업데이트

dye0p committed:
refactor: ADMIN 권한이 있는 사람만 admin 관련 페이지 진입이 가능하도록 리팩터링

- 사용자와 관리자의 예약 방법을 다르게 하기 위해서  요청의 쿠키의 권한을 인터셉터로 검증하여 관리자 권한을 가진 사람만 /admin/** 페이지에 접근할 수 있도록 하였음

- member의 권한을 분류하기 위해서 기존의 basic으로 고정되어있던 role을 Enum 클래스로 별도로 관리하도록 변경

dye0p committed:
feat: 예약자별, 테마별, 날짜별 검색 조건을 사용해 예약을 검색 하는 API 구현

dye0p committed:
refactor: 패키지 구조화

dye0p committed:
refactor: 사용자를 찾을 수 없을 때 예외 메세지 수정

- 예외 메세지 사용자 친화적으로 수정

dye0p committed:
refactor: 재할당 불가 변수 final 키워드 추가

dye0p committed:
docs: 기능 구현 완료사항 업데이트

dye0p commented on src/main/java/roomescape/service/reservation/ReservationService.java:
```diff
@@ -0,0 +1,156 @@
+        return reservationDao.existsReservationBy(date, timeId, themeId);
+    }
+
+    public List<ReservationResponseDto> findByCondition(final SearchConditionRequest request) {
+        final Long themeId = request.themeId();
+        final Long memberId = request.memberId();
+        final LocalDate dateFrom = request.dateFrom();
+        final LocalDate dateTo = request.dateTo();
+
+        final List<Reservation> reservations = reservationDao.findByDate(dateFrom, dateTo);
```
**[설계 판단에 대한 질문]**

날짜 범위는 DAO 계층에서 WHERE 조건으로 필터링하고,
나머지 조건(회원 ID, 테마 ID 등)은 서비스 계층에서 처리하는 방식을 선택했습니다. 

이유는 아래와 같습니다.
1. 모든 조건을 쿼리로 처리할 경우, 비즈니스 로직이 DAO에 결합되어 유지보수에 불리하다고 판단했습니다.
2. 서비스 계층에서 조건 필터링을 처리하면, 테스트 용이성과 비즈니스 책임 분리가 더 명확해진다고 보았습니다.
3. 다만, 전체 데이터를 다 가져오는 것은 성능상 부하가 크므로, 기간 조건만은 DB에서 필터링하여 최소 범위만 가져오도록 했습니다.

하지만 구현을 다 하고, 조금 더 생각 해봤을 때 해당 검색의 조건이 **서비스에서 처리해야할 비즈니스 로직인가?** 라는 생각이 들었습니다. 요청으로 들어온 조건은 db에 존재하는 데이터들로 필터링을 하는 것이 더 나은 방법이라고 느껴졌습니다. 추가로 성능 부하때문에 기간으로 필터링을 한다고 해도, 데이터가 많은 경우 이것 역시 성능에 부하가 갈 것이고, 모든 조건으로 필터링을 하는 것이 더 나은 선택이라고 판단되었습니다. 

이렇게 구현을 한 배경에는 service 계층에서 수행해야할 로직과 dao 계층에서 수행해야할 로직의 판단 기준이 명확하지 않아서라고 생각합니다.

아직 코드를 변경하지 않았지만 지금과 같은 생각의 흐름과 코드에 대한 이스트의 생각이 궁금합니다!

EastHShin replied:
날짜 조건만 필터링해 관련없는 데이터를 모두 가져오는 방식으록 구현한다면, 데이터가 정말 많아졌을때는 어떨까요?
날짜 범위에 해당하는 데이터를 가져오는 것만으로 DB에 대한 부하와 데이터를 가져왔을때 메모리 부족 현상이 있을 것 같네요~

필터링 하는 로직을 잘 수행할 수 있는 것은 DB 이고, 서비스 계층에서는 DB 계층에 쿼리 조건을 잘 넘겨주기 위한 책임만 가져도 좋을 것 같아요. 현재는 조건을 그대로 dao 로 전달해 서비스계층에서 하는일이 없다고 느껴질 수 있지만, 복잡한 검색 조건이 들어오게 된다면 이러한 검색 쿼리를 가공하는 것 또한 많은 로직이 필요해 서비스 계층에서 할 일이 많아질거에요 :)

dye0p replied:
이스트의 의견에 동의합니다! 지금은 데이터가 적어, 지금처럼 필터링 하는 방식이 간단하고 좋아보이지만, 추후에 데이터가 정말 많아진다면 필요하지 않은 리소스를 조회하면서 성능의 문제가 발생할 것 입니다. 그리고 조건에 따라 필터링의 조건은 DB에서 수행해야할 책임이고, 그 외의 로직은 비즈니스 로직으로 처리해야 한다고 생각합니다! 모든 조건을 쿼리문으로 필터링 하도록 수정하였습니다!

[refactor: 예약자, 날짜, 테마 검색 로직 변경](https://github.com/dye0p/spring-roomescape-member/commit/3baa06a45ff54449b698f5f5dc8f3a5b084b61db)

dye0p commented on src/test/java/roomescape/service/reservation/ReservationServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+import roomescape.dto.reservationtime.response.ReservationTimeResponseDto;
+import roomescape.dto.theme.response.ThemeResponseDto;
+
+public class ReservationServiceTest {
+
+    private static final LocalDate NOW_DATE = LocalDate.now().plusDays(1);
+
+    private ReservationService reservationService;
+    private ReservationDao reservationDao;
+    private ReservationTimeDao reservationTimeDao;
+    private ThemeDao themeDao;
+    private MemberDao memberDao;
+
+    @BeforeEach
+    void init() {
+        reservationDao = new FakeReservationDaoImpl();
+        reservationTimeDao = new FakeReservationTimeDaoImpl();
+        themeDao = new FakeThemeDaoImpl();
+        memberDao = new FakeMemberDaoImpl();
+        reservationService = new ReservationService(reservationDao, reservationTimeDao, themeDao, memberDao);
+    }
```
**[Service 테스트에서 Fake 객체를 사용해야 하는가]**

현재 `Service` 계층 테스트에서 `FakeDao`를 활용하고 있는데, 테스트를 작성하다 보니 **꼭 이렇게까지 해야 할까?**, 혹은 **이 선택이 정말 더 나은 걸까?** 하는 의문이 들었습니다. 그래서 아래와 같이 각각의 방식에 대한 제 나름의 생각과 장단점을 정리해봤고, 이에 대한 이스트의 의견도 궁금합니다!

**FakeDao 사용**
장점
- Service라는 테스트의 독립성을 지킬 수 있다 (db에 영향을 받지않고, 오직 단일 테스트에 집중할 수 있음)
- 실제 db를 사용하지 않기 때문에 Spring 컨테이너를 띄울 필요가 없어, 속도가 빠르다 (테스트가 많아질수록 이득)

단점
- 매번 FakeDao 객체를 구현해주어야 한다.
    - Service 객체에 의존성 주입을 해줘야 하기 때문에 테스트용  configuration도 설정해줘야 한다. 
- FakeDao는 개발자가 직접 구현하기 때문에 실수를 하게 되면 실제 dao와 다르게 동작할 수 있다.
   - FakeDao도 테스트 대상이 되는지 의문이다. 
- 위의 단점은 Service 테스트의 잘못된 결과로 이어질 수 있다.

**프로덕션 dao 사용**
장점
- 프로덕션 dao를 사용하기 때문에 코드가 잘못 동작하지 않는다.
- 만약 잘못 동작한다면 자연스럽게 dao의 검증도 가능하다고 생각한다(이건 장점이자 단점이라고 생각하는 개인적인 의견인데, 틀렸다면 지적해주세요!)
- FakeDao 객체를 만들지 않아도 된다.

단점
- 실제 dao를 사용하기 때문에 Spring 컨테이너를 띄워야하므로 테스트 속도가 느리다(테스트가 많아질수록 더 느리다)
- 테스트가 실패한다면 Service에서 실패한 테스트인지, dao에서 실패한 테스트인지 구분하기가 어렵다
   - dao에서 실패한다면 Service 테스트가 깨져버려, 정상동작하는 Service 코드도 테스트가 안되는 이상한 상황이 발생한다. 

**결론 (하지만 아직 확신은 없습니다…)**
Service 테스트는 최대한 빠르게, 독립적으로 유지하는 것이 좋겠다고 판단하여 현재는 FakeDao를 사용하는것이 좋겠다고 판단하였습니다. Fake 객체를 사용하는 테스트 전략은 어떤 상황에서 더 적절하며, **어떤 경우에는 오히려 해가 될 수도 있는가?** 
에 대한 이스트의 생각이 궁금합니다!

EastHShin replied:
통합테스트를 진행하느냐, 서비스 계층의 단위테스트를 진행하느냐에 대한 고민이신 것 같아요~
>프로덕션 dao 사용
장점
만약 잘못 동작한다면 자연스럽게 dao의 검증도 가능하다고 생각한다(이건 장점이자 단점이라고 생각하는 개인적인 의견인데, 틀렸다면 지적해주세요!)

이 부분은 반 정도는 맞다고 할 수 있는데요ㅎㅎ 물론 DAO 의 동작이 검증 가능한 것도 사실이지만 DAO 의 동작이 정상적으로 동작하는지 100% 신뢰할 수는 없다고 생각해요. DAO 의 동작이 잘못되어 정상적이지 않은 쿼리가 발생하더라도 예외가 발생하지 않는다면 정상 실행되고, 비즈니스 로직 수행상 문제없이 테스트코드가 통과될 수 있기 때문에 DAO 를 완벽히 검증할수는 없다고 생각해요. 그렇기 때문에 계층별 단위 테스트가 필요한 것이구요~

그리고 통합테스트와 단위테스트는 모두 필요하다고 생각해요. 이 둘의 목적이 다르기 떄문이죠. (이 둘의 목적은 정리해서 코멘트 남겨주시면 좋을 것 같습니다ㅎㅎ)

그리고 FakeDAO를 직접 구현하는게 번거롭고, 실수할 여지도 많다고 하셨는데, 이러한 불편함을 해소하기 위해 mocking 도구들이 있습니다. Mockito 에 대해 한번 알아보시면 좋을 것 같아요.
Mockito 를 처음부터 사용하는 것보다 지금처럼 직접 불편을 겪어보고 필요에 의해 도구를 찾게 되는 것이 의미있는 과정이라고 생각되네요ㅎㅎ

dye0p replied:
피드백을 보고 아직 테스트에 대한 지식이 부족함을 느꼈습니다.

말씀해주신 단위 테스트와 통합 테스트의 목적을 정리하고 개인적인 의견을 남겼습니다!

### 단위 테스트
각각의 단위에 대해서 검증을 하기 때문에 작은 단위를 검증하기 위해서 필요하다.
외부 시스템과 의존하지 않는 환경에서 테스트를 실행하기 때문에 해당 단위에 대한 정확한 검증이 가능하다.

### 통합 테스트
웹 환경과 같은 외부 시스템과 의존을 하게되는 경우 요청에 대한 응답이 의도한 대로 동작하는지 검증이 필요하다.
통합 테스트는 여러 컴포넌트들의 구성이 서로 잘 통신하는지 검증해야 한다. 외부의 환경을 포함하여 검증하기 때문에 실제 환경과 비슷한 테스트가 가능하다.

아직 통합 테스트라는 의미가 정확하게 와닿지 않는 것 같습니다. 하지만 스프링과 같은 웹 개발 환경에서 통합 테스트는 필수라고 생각합니다. 피드백에서 제시해주신 다양한 mocking 도구들에 대해서 공부해보면 좋을 것 같습니다! 감사합니다 : )

dye0p reviewed (commented):
안녕하세요 이스트! 질문이 필요한 코드에 대해서 코멘트를 남겼습니다! 리뷰 잘 부탁드립니다 : )

EastHShin commented on build.gradle:
```diff
@@ -25,6 +25,11 @@ dependencies {
+    // JWT 의존성
+    implementation 'io.jsonwebtoken:jjwt-api:0.11.5'
+    runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.11.5'
+    runtimeOnly 'io.jsonwebtoken:jjwt-jackson:0.11.5'
```
공통적으로 사용되는 버전 정보는 변수로 추출해 관리해볼까요?

dye0p replied:
의존성도 변수화를 할 수 있는지 몰랐습니다..!  의존성을 변수화 하면 어떤 장점이 있는지 궁금해 찾아봤습니다.

- 의존성이 많을때 같은 버전을 사용하는 의존성이 흩어져 있을 때 한곳으로 모아 관리할 수 있다.
- 의존성의 버전을 바꿔야 할때 한번에 바꿀 수 있다.
- 의존성이 많아지고 복잡해질때 의존성을 변수화 하면 의도치 않은 버전 불일치를 예방할 수 있다.

생각보다 장점이 많다는 것을 알 수 있었습니다! 감사합니다👍

[chore: jwt 의존성 변수화](https://github.com/dye0p/spring-roomescape-member/commit/27828ed1097c2863e9d380d98e7f77f839c529bc)

EastHShin commented on src/main/java/roomescape/dto/member/request/MemberRegisterRequest.java:
```diff
@@ -0,0 +1,21 @@
+        String name,
+
+        @NotBlank(message = "이메일은 필수 입력입니다.")
+        String email,
```
이메일 형식이 아니어도 가입을 할 수 있군요? 이메일 형식인지를 요청에서 걸러줄 수 있으면 좋을 것 같아요~

dye0p replied:
회원가입 요청 dto에서 이메일 형식을 검증할 수 있도록 수정하였습니다!

[feat: 회원가입 요청 이메일 형식 검증 추가](https://github.com/dye0p/spring-roomescape-member/commit/378a8e1a19de56360a45b75d277fe7e304c1b27b)

EastHShin commented on src/main/java/roomescape/common/exception/GlobalExceptionHandler.java:
```diff
@@ -33,4 +38,15 @@ public ResponseEntity<String> handleMethodArgumentNotValidException(MethodArgume
+    @ExceptionHandler(InvalidAuthException.class)
+    public ResponseEntity<String> handleMemberException(InvalidAuthException e) {
+        return new ResponseEntity<>(e.getMessage(), HttpStatus.UNAUTHORIZED);
+    }
```
401, 403의 차이가 무엇인가요? 이에 대해 학습해보고 예외 핸들링을 구분해보면 좋겠습니다 :)

dye0p replied:
`401 Unauthorized` 상태코드는 클라이언트의 인증이 유효하지 않은 경우 응답하는 상태코드 입니다.
- 클라이언트의 로그인이 유효하지 않은 경우
- 로그인 인증정보를 찾을 수 없는 경우

`403 Forbidden` 상태코드는 인증된 클라이언트가 특정 URL에 접근하거나 특정 권한이 없는 경우 응답하는 상태코드 입니다.
- /admin/**에 접근할 수 있는 사용자는 ADMIN권한을 가진 사용자만 접근할 수 있다.
- ADMIN 권한을 가지지 않는 사용자는 접근할 수 없다

인증과 인가의 차이를 제대로 알지 못했습니다. 인증에 대한 예외는 401로 응답, 인가에 대한 예외는 403 상태코드로 응답하도록 수정하였습니다!

[refactor: 인가 권한이 없는 경우 처리 예외 변경](https://github.com/dye0p/spring-roomescape-member/commit/3cb5949558c5a74f77fb8306548aa96dbf0a733e)

EastHShin commented on build.gradle:
```diff
@@ -25,6 +25,11 @@ dependencies {
 
+    // JWT 의존성
+    implementation 'io.jsonwebtoken:jjwt-api:0.11.5'
+    runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.11.5'
```
implementation, runtimeOnly 의 차이는 무엇인가요?

dye0p replied:
`implementation`, `runtimeOnly` 의 가장 큰 차이점은 의존성의 사용 시점 입니다.

implementation 을 사용한 의존성은 컴파일 시점과 런타임 시점에서 모두 해당 의존성을 사용할 수 있습니다. 하지만 runtimeOnly 의존성은 런타임 시점에만 의존성을 사용할 수 있습니다. 컴파일 시점에서 의존성을 사용한다는 것은 우리가 실제로 코드를 작성할 때 사용되는 의존성을 얘기합니다. 

예를 들어 
``` java
    public String createToken(final Member member) {
        return Jwts.builder()
                .setSubject(member.getId().toString())
                .claim("name", member.getName())
                .claim("role", member.getRole())
                .signWith(Keys.hmacShaKeyFor(SECRET_KEY.getBytes()))
                .compact();
    }
```
위 코드를 작성하고 빌드하기 위해서는 클래스, 메서드가 필요합니다. 따라서 `Jwts.builder()`, `signWith()` 를 사용하기 위해서는 jwt 의존성이 필요합니다. 이 때 implementation 를 통해서 jwt 의존성을 추가해주면 컴파일 시점에서 해당 의존성을 사용할 수 있습니다. 

runtimeOnly 를 사용한 의존성은 런타임 시점에서 동작하는 의존성 입니다. 여기서 런타임은 빌드를 마치고 프로그램이 실행되는 시점 입니다. 런타임에 실행되는 코드는 프로그램 내부에서 보통 구현체가 실행되는 시점 입니다. 런타임에 동적으로 실행되기 때문에 개발자가 코드를 직접 작성하지 않습니다.

위 createToken 메서드를 작성하고 실행하면 `.signWith()`의 내부 동작은 런타임에 이루어 집니다. 이 때 내부적으로 하는 동작은 runtimeOnly로 설정된 의존성에서 처리합니다.

**🤔 왜 둘로 나누었을까?**
런타임 시점에 필요한 의존성을 implementation 의존성으로 설정하면 컴파일 단계에서 필요하지 않은 의존성이 추가됩니다. 의존성이 많아질수록 빌드 속도가 저하 됩니다.

**🤔 같은 의존성인데 왜 방식에 따라서 빌드 속도가 차이가 나지?**
Gradle은 컴파일 단계에서 runtimeOnly 의존성을 무시합니다. 하지만 implementation로 지정된 의존성은 컴파일 단계에서 모두 빌드합니다. 그렇기 때문에 필요하지 않은 시점의 의존성을 implementation로 추가하는것은 리소스 낭비 입니다.

결론적으로 어떠한 상황에서 사용되는 의존성인지 파악을 하고, 불필요한 리소스를 낭비하지 않기 위해서 적절한 방식으로 의존성을 추가해야 합니다.

EastHShin commented on src/main/java/roomescape/controller/admin/AdminController.java:
```diff
@@ -0,0 +1,41 @@
+        return ResponseEntity.created(URI.create("/admin/reservations" + response.id())).body(response);
+    }
+
+    @GetMapping("/search/")
```
끝에 `/` 는 왜 붙여주셨나요? 꼭 필요한가요?

dye0p replied:
실수로 붙인것 같아요.. 수정했습니다!

[refactor: 불필요한 URL 경로 수정](https://github.com/dye0p/spring-roomescape-member/commit/7a41009bebcf7af609746eb09ee1a642b3111146)

EastHShin commented on src/main/java/roomescape/controller/auth/AuthController.java:
```diff
@@ -0,0 +1,58 @@
+        return ResponseEntity.ok(response);
+    }
+
+    @PostMapping("/logout")
+    public ResponseEntity<Void> logout(final HttpServletResponse response) {
+        final Cookie expireCookie = cookieProvider.createExpireCookie(COOKIE_NAME);
```
로그아웃도 구현해주셨군요 👍

dye0p replied:
요구사항엔 없었지만 로그인이 있다면 회원가입도 있어야 한다고 판단했습니다 : )

EastHShin reviewed (changes requested):
안녕하세요 도기~

2단계도 잘 구현해주셨네요!
질문에 대한 답변과 코멘트 남겨두었으니 확인 부탁드려요 :)

dye0p committed:
chore: jwt 의존성 변수화

dye0p committed:
refactor: 불필요한 URL 경로 수정

dye0p committed:
feat: 회원가입 요청 이메일 형식 검증 추가

dye0p committed:
refactor: 인가 권한이 없는 경우 처리 예외 변경

- 인가 권한이 없는 경우 403 상태코드를 응답할 수 있도록 커스텀 예외 추가

dye0p committed:
refactor: 예약자, 날짜, 테마 검색 로직 변경

- 기존 날짜만으로 필터링 하고, 서비스 로직에서 테마와 사용자를 필터링 하던 로직은 데이터가 많아지면 성능에 좋지않음
- 필터링의 조건은 db가 수행해야 하는 책임이라 판단하고, 모든 조건을 쿼리로 필터링 하도록 수정하였음

dye0p reviewed (commented):
안녕하세요 지노! 말씀하신 피드백을 반영하였습니다! 리뷰 잘 부탁드립니다 : )

EastHShin commented on src/main/java/roomescape/domain/member/Member.java:
```diff
@@ -0,0 +1,83 @@
+        return new Member(id, name, email, password, role);
+    }
+
+    public void validatePassword(final String password) {
```
비밀번호가 평문으로 저장 및 비교되고 있는데, 보안적 측면을 고려해 개선해볼 수 있을까요?

dye0p replied:
평소 암호화에 대한 지식이 부족해, 자바와 스프링 환경에서 비밀번호를 어떻게 암호화 할 수 있는지 학습해보았습니다. 

암호화 방향은 양방향과 단방향 암호화중 단방향 암호화를 하였습니다.
양방향 암호화는 복호화가 가능하다는 특징이 있고, 단방향 암호화는 복호화가 불가능하다는 특징이 있습니다. 
비밀번호와 같은 정보는 탈취당했을 때 복호화가 가능하면 치명적일 수 있기 때문에 단방향 암호화를 선택했습니다.

단방향 암호화는 해쉬 알고리즘을 사용합니다. 이를 자바와 스프링에서 구현하는 방법에는 자바에서 기본으로 제공하는 `MessageDigest` 를 사용하는 것과, 스프링 환경에서는 `Spring Security`로 `BCrypt`를 사용한 암호화가 가능합니다.

해당 미션에는 `MessageDigest` 방식을 사용하였는데, 이유는 다음과 같습니다.
- 왜 `Spring Security`가 많이 사용되는지 `MessageDigest`를 사용하면서 느껴보고 싶었습니다.

지금은 학습의 차원에서 `MessageDigest`를 사용하였지만, `salt`를 지원하지 않는점과, 보안적인 부분에서 안전하지 않다는 의견이 많아서 ` Spring Security`를 도입하는 것도 좋을 것 같습니다. 

평소 암호화와 보안에 대한 지식이 미흡했는데, 이번 기회를 통해서 조금이나마 알게 되었습니다. 감사합니다!

혹시 저의 코멘트에서 틀린 부분이 있다면 피드백 해주시면 감사하겠습니다 : )

[feat: 비밀번호 암호화 기능 추가](https://github.com/dye0p/spring-roomescape-member/commit/df3bd754a967c8309b5074459eb1cfba8b939ebf)

EastHShin replied:
네 현재는 구현해주신 것으로 충분합니다 :) spring security는 추후에 공부해보면 좋겠네요

EastHShin commented on src/main/java/roomescape/dto/reservation/request/AvailableReservationTimeRequestDto.java:
```diff
@@ -1,12 +1,11 @@
 public record AvailableReservationTimeRequestDto(
 
-        @NotBlank(message = "날짜는 필수입니다.")
+        @NotNull(message = "날짜는 필수입니다.")
```
`@NotNull` 로 바꾸어주신 이유가 있나요? `@NotBlank` 와는 어떤 차이가 있나요?

dye0p replied:
`@NotBlank`는 `null`을 포함하여 `""`, `" "` 빈 공백 검증도 함께 합니다. 빈 공백이 올 수 있는 타입은 String 타입입니다. 
따라서 `@NotBlank`가 지원하는 타입은 `String` 타입 입니다. 
만약 String 타입이 아닌 다른 타입에 사용한다면 `no validator could be found for constraint` 예외가 발생할 수 있습니다.

Strnig 타입의 필드를 제외한 다른 타입의 필드엔 `@NotNull`을 사용해야 합니다. 
`@NotNull`은 `null`을 허용하지 않고, `""`, `" "` 은 허용합니다.
`@NotNull`은 `String` 을 포함한 모든 타입에서 사용할 수 있습니다.

정리하자면 `LocalDate` 타입의 필드는 `@NotBlank` 를 사용할 수 없고, `@NotNull`을 사용해야 합니다!

아래에서 각각의 특징을 정리해봤습니다!

### `@NotBlank`

- 필드에 `null`을 포함한 공백도 허용하지 않는다.
- 허용되지 않는 것
  - `null`
  - `””`, `“ “`
- String 타입에 대해서 지원한다.

### `@NotNull`
- 필드에 `null` 을 허용하지 않는다.
- 공백은 허용한다. 즉, 필드가 비어있을 수 있다.
- 허용되지 않는 것
  - `null`
- 허용되는 것
  - `””`, `“ “`
- `String` 뿐만 아니라 `int`, `long` 등등 다 검사가 가능하다.
- `Integer`, `String` 등 원시 타입이 아닌 래퍼 타입도 지원한다.

EastHShin commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -25,28 +25,28 @@ void setup() {
-                .body("size()",
-                        is(0)); // 아직 생성 요청이 없으니 Controller에서 임의로 넣어준 Reservation 갯수 만큼 검증하거나 0개임을 확인하세요.
-    }
+//    @Test
+//    void 일단계() {
+//        RestAssured.given().log().all()
+//                .when().get("/admin")
+//                .then().log().all()
+//                .statusCode(200);
+//    }
+//
+//    @Test
+//    void 이단계() {
+//        RestAssured.given().log().all()
```
주석처리해 주신 이유가 있을까요?

dye0p replied:
테스트 과정에서 실패하는 테스트를 우선 주석처리 하고 추후에 수정하려 했는데, 잊어버린 것 같습니다..! 다시 봤을 때 해당 테스트는 다른 테스트에서 이미 수행중인 테스트여서 중복테스트라 생각해, 제거하였습니다!

[remove: 중복되는 테스트 제거](https://github.com/dye0p/spring-roomescape-member/commit/f921c6b0d6603278038ed6f62dc710f43cca3f6b)

EastHShin commented on src/main/java/roomescape/common/auth/JwtTokenProvider.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public Long getSubjectFromPayloadBy(final String token) {
+        return Long.valueOf(Jwts.parserBuilder()
+                .setSigningKey(Keys.hmacShaKeyFor(SECRET_KEY.getBytes()))
+                .build()
+                .parseClaimsJws(token)
```
parseClaimsJws() 를 진행하면서 다양한 예외가 발생할 수 있는데, 이에 대한 예외도 핸들링해보면 어떨까요?

dye0p replied:
토큰 파싱 과정에서 예외가 발생할 것이란 생각을 못했습니다! `parseClaimsJws()` 메서드를 보니 예외를 호출한 곳에서 처리하도록 던져주고 있는것을 확인했습니다. 그래서 해당 메서드를 호출하는 곳에서 예외처리를 해주었습니다!

[feat: 토큰 파싱 과정 예외처리 추가](https://github.com/dye0p/spring-roomescape-member/commit/3b1971c03cab88c21ad9c1ed1b3f85ac30c4ad03)
[refactor: 상수화 되어있던 토큰 시크릿 키 properties 로 이동](https://github.com/dye0p/spring-roomescape-member/commit/ff87216d1fe1f5b00cd1876ffa437accaf56e00f)
[feat: 토큰 커스텀 예외 추가](https://github.com/dye0p/spring-roomescape-member/commit/593a2ed8e149bde43bf709fce617cbdad9abed4c)

EastHShin reviewed (changes requested):
안녕하세요 도기~

피드백 반영 잘 해주셨네요!
몇가지 추가 코멘트 남겼으니 확인 부탁드려요 :)

dye0p committed:
remove: 중복되는 테스트 제거

- controller 테스트를 통해서 해당 테스트를 진행해주고 있으므로 중복되는 테스트를 제거함

dye0p committed:
feat: 토큰 파싱 과정 예외처리 추가

dye0p committed:
refactor: 상수화 되어있던 토큰 시크릿 키 properties 로 이동

dye0p committed:
feat: 토큰 커스텀 예외 추가

dye0p committed:
feat: 쿠키 null 검증 추가

dye0p committed:
refactor: row로 member를 생성할 때 비밀번호를 포함하지 않도록 변경

- db에서 비밀번호를 포함하여 member를 생성하여 비밀번호를 메모리에 올리는것은 좋지않기 때문에 변경

dye0p committed:
feat: 비밀번호 암호화 기능 추가

dye0p reviewed (commented):
안녕하세요 이스트! 피드백에 대해 수정을 하여 다시 리뷰요청 합니다! 감사합니다 : )

EastHShin reviewed (approved):
안녕하세요 도기~

미션 진행하시느라 고생하셨습니다!
남은 우테코 기간도 화이팅하세요 :)
